Mathlib/Probability/Hoeffding.lean

```diff
@@ -0,0 +1,233 @@
+/-
+Copyright (c) 2024 Kei Tsukamoto. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kei Tsukamoto, Kazumi Kasaura, Yuma Mizuno, Naoto Onda, Sho Sonoda
+-/
+
+import Mathlib.Probability.Moments.Basic
+import Mathlib.Probability.Moments.MGFAnalytic
+import Mathlib.Probability.Variance
+import Mathlib.Analysis.Calculus.ParametricIntegral
+import Mathlib.Analysis.SpecialFunctions.Log.Deriv
+import Mathlib.MeasureTheory.Measure.Tilted
+import Mathlib.Analysis.Calculus.Taylor
+
+/-!
+# Hoeffding's lemma
+
+This file states Hoeffding's lemma.
+
+## Main results
+
+* `ProbabilityTheory.hoeffding`: Hoeffding's Lemma states that for a random variable `X` with
+  `E[X] = 0` (zero mean) and `a ≤ X ≤ b` almost surely, the inequality
+  `mgf X μ t ≤ exp (t^2 * (b - a)^2 / 8)` holds almost surely for all `t ∈ ℝ`.
```

Comment:
It would be nice to prove the more general theorem that a bounded random variable (with no mean zero assumption) is subgaussian (rather than having subgaussian MGF)

Reply:
To "remove" the mean zero assumption it's possible to write an immediate corollary of the current result: if `X` is bounded in an interval then `X - E[X]` is also bounded in an interval (of same length as `X`), so has sub-Gaussian moment generating function.

If you are thinking about proving one of the other properties called sub-Gaussian, then please tell us which one! And that's probably for another PR to avoid scope creep on this one.